File: app/domain/shuffle_order.py
"""Shuffle ordering, extracted from StreamEngine.set_shuffle_enabled.

Pure: the RNG is injected so tests are deterministic.
"""

from __future__ import annotations

import random
from collections.abc import Sequence


def shuffled_order(ids: Sequence[int], rng: random.Random | None = None) -> list[int]:
    order = list(ids)
    (rng or random).shuffle(order)
    return order


def restore_order(current: Sequence[int], restore: Sequence[int] | None) -> list[int] | None:
    """Return the canonical order for `current` ids given a saved `restore`
    snapshot; None when nothing to restore (ids vanished from the queue)."""
    if not restore:
        return None
    known = set(current)
    ordered = [i for i in restore if i in known]
    return ordered or None

File: packages/domain/src/shuffle-order.ts
/** Shuffle ordering. Ported from app/domain/shuffle_order.py. RNG injected. */

export function shuffledOrder(ids: readonly number[], rng: { shuffle(list: number[]): void }): number[] {
  const order = [...ids];
  rng.shuffle(order);
  return order;
}

/**
 * Canonical order for `current` ids given a saved `restore` snapshot;
 * null when nothing to restore (all saved ids vanished).
 */
export function restoreOrder(
  current: readonly number[],
  restore: readonly number[] | null,
): number[] | null {
  if (!restore || restore.length === 0) return null;
  const known = new Set(current);
  const ordered = restore.filter((id) => known.has(id));
  return ordered.length > 0 ? ordered : null;
}
